### PR TITLE
[RFC] Pytest environment improvements

### PR DESCRIPTION
**Pull request by u99127**
In many places having a global pytest flag is useful . For me with the
build and test of tvm , I would like to be able to globally pass in
pytest options as part of development flow or CI flows where one would
like to measure other things regularly that need measurements including
pytest coverage data that I would like to experiment with across the stack.
    
This has been achieved with an additional setup-pytest-env.sh file in
tests/scripts rather than putting in something in every single task test
script and something I would like to avoid.
    
This now means the -v option to pytest is superfluous. I did consider
having a pytest.ini file but that doesn't allow me to pass any old
environment variable in and this seems to be the best compromise.

I have also updated the pull request documentation to reflect
the presence of the -k pytest option at the same time though
technically this could be a separate pull request.

@tqchen

### CODE REVIEW

**Comment by u99127** [comment]
I don't like my current hack of overloading "with_same_user" for sourcing this global environment but it seemed like the simplest hack and worked in my environment. Obviously I don't have cuda testing in my CI or my regular test environment, so this isn't fully clear.

**Comment by tqchen** [comment]
It would be great if we can avoid the hack into the `with_same_user`. One alternative would be still pass in the `PYTEST_ADDOPTS` env variable from the docker env(for development purposes) but source the setup-pytest-env within each of the script.

This also makes the intention of the CI scripts clear, since sometimes we could directly use these scripts. For example, I sometimes would use `./tests/scripts/task_python_unittest.sh` without the docker image to run the testcases

**Comment by u99127** [comment]
> 
>

Thanks for the quick review.

> It would be great if we can avoid the hack into the `with_same_user`. One alternative would be still pass in the `PYTEST_ADDOPTS` env variable from the docker env(for development purposes) but source the setup-pytest-env within each of the script.
> 

Indeed , that seems a simple enough solution, here's a rework. 

> This also makes the intention of the CI scripts clear, since sometimes we could directly use these scripts. For example, I sometimes would use `./tests/scripts/task_python_unittest.sh` without the docker image to run the testcases

Yeah our wrapping on top uses docker most of the time and that's why I went with the "with_same_user" hack for a prototype but I can see that others may not choose to develop in the same way as us. 

R

**Comment by tom-gall** [comment]
> It would be great if we can avoid the hack into the `with_same_user`. One alternative would be still pass in the `PYTEST_ADDOPTS` env variable from the docker env(for development purposes) but source the setup-pytest-env within each of the script.
> 
> This also makes the intention of the CI scripts clear, since sometimes we could directly use these scripts. For example, I sometimes would use `./tests/scripts/task_python_unittest.sh` without the docker image to run the testcases

Exactly. For my modest arm dev board environments, where i want/need to run tests, docker just isn't viable.

**Comment by u99127** [comment]
> 
> 
> > It would be great if we can avoid the hack into the `with_same_user`. One alternative would be still pass in the `PYTEST_ADDOPTS` env variable from the docker env(for development purposes) but source the setup-pytest-env within each of the script.
> > This also makes the intention of the CI scripts clear, since sometimes we could directly use these scripts. For example, I sometimes would use `./tests/scripts/task_python_unittest.sh` without the docker image to run the testcases
> 
> Exactly. For my modest arm dev board environments, where i want/need to run tests, docker just isn't viable.

That is certainly a very interesting orthogonal question as we would need to figure out how best to run things cross as not every single python package tvm depends on is available natively on AArch32 or AArch64.

regards
Ramana

File: docs/contribute/pull_request.rst [hunk 1]
```diff
@@ -118,3 +118,6 @@ If you want to run a single test:
   rm -rf python/tvm/*.pyc python/tvm/*/*.pyc python/tvm/*/*/*.pyc
 
   TVM_FFI=ctypes python -m pytest -v tests/python/unittest/test_pass_storage_rewrite.py
+
+  #Additionally if you want to run a single test, for example test_all_elemwise inside a file.
```
Comment (tqchen) [inline docs/contribute/pull_request.rst:122]: not : space between `#` and `Additionally`

Comment (tqchen) [reply docs/contribute/pull_request.rst:122]: ` # Additionally if`

Comment (u99127) [reply docs/contribute/pull_request.rst]: Ah, thanks - fixed.

**Comment by tqchen** [comment]
Thanks @u99127 @tom-gall

**Comment by roastduck** [comment]
I have a question. Actually I cannot directly run `./tests/scripts/task_python_unittest.sh` without Docker after merging this PR.

```
tests/scripts/setup-pytest-env.sh: line 23: CI_PYTEST_ADD_OPTIONS: unbound variable
```

What is `CI_PYTEST_ADD_OPTIONS`? Should I set it to an empty string?

**Comment by tqchen** [comment]
hmm, @u99127 do you mind send a patch? we should detect if CI_PYTEST_ADD_OPTIONS is available and set it to none

**Pull request merged**